### PR TITLE
Generate composition grading queues

### LINKED ISSUE #441: Web interface to create queues for composition grading
I thought this button existed at one point, but couldn't find it just now... I think a natural place for such a button would be the staff-facing assignment-edit page.

### PR DESCRIPTION
**Pull request by alvinwan**
fixes #441 

basic checklist/'*' input:
generate-basic-queue

If you'd like to assign all final submissions to two misbehaving readers:
generate-csv-queue

old:
queue-generation

@Sumukh @soumyabasu Ready. This took a bit more code, so I'll wait until you guys give the green light, to merge.

### CODE REVIEW

**Comment by alvinwan** [comment]
(for the wiki, ignore)

adin-tutorial-assignments
adin-tutorial-assignments2
adin-tutorial-assignments3

File: server/app/api.py [hunk 1]
```diff
@@ -439,6 +439,18 @@ def enrollment(self):
                 })
         return json.dumps(data)
 
+    def check(self, emails, course, role):
```
Comment (Sumukh) [inline server/app/api.py]: What is this used for? The ok-client enrollment?

File: server/app/api.py [hunk 2]
```diff
@@ -1987,6 +2010,72 @@ def new_entity(self, attributes):
...
+            subm.queue = queues[i].key
+            subm.put()
+            i = (i + 1) % len(staff)
+            
+        return queues
+
+    def check_permissions(self, user, data):
+        course = data['course'].get()
+
+        if user.key not in course.staff and not user.is_admin:
```
Comment (Sumukh) [inline server/app/api.py]: This method should probably return True/False and leave it up to the caller what to do.

File: server/app/api.py [hunk 3]
```diff
@@ -1987,6 +2010,72 @@ def new_entity(self, attributes):
...
+                'course': KeyArg('Course', required=True),
+                'assignment': KeyArg('Assignment', required=True)
+            }
+        }
+    }
+    
+    def generate(self, user, data):
+        """ Splits up submissions among staff members """
+        
+        self.check_permissions(user, data)
```
Comment (Sumukh) [inline server/app/api.py]: See check_permission comment.

File: server/app/api.py [hunk 4]
```diff
@@ -1987,6 +2010,72 @@ def new_entity(self, attributes):
...
+        course, assignment = data['course'].get(), data['assignment'].get()
+        userify = lambda parts: [part.user.get() for part in parts]
+        
+        staff = userify(models.Participant.query(
```
Comment (Sumukh) [inline server/app/api.py]: We don't necessarily want all staff grading composition. We'd need a selection of staff. The front-end could display a list of staff and the user could select the users to choose. 

Perhaps the create queues button goes a page `/assignId/queues/new` choosing what staff members get assigned. That page can then include the userIDs selected in it's post.

File: server/app/api.py [hunk 5]
```diff
@@ -1987,6 +2010,72 @@ def new_entity(self, attributes):
...
+            models.Participant.role == STAFF_ROLE,
+            models.Participant.course == course.key).fetch())
+
+        students = userify(models.Participant.query(
+            models.Participant.role == STUDENT_ROLE,
+            models.Participant.course == course.key).fetch())
+
+        subms, queues = [], []
+
+        for student in students:
```
Comment (Sumukh) [inline server/app/api.py]: Wouldn't this double submit into a queue if a student is part of a group? I think it'd be better to pull from FinalSubmissions.

File: server/static/js/admin/controllers.js [hunk 6]
```diff
@@ -223,6 +223,92 @@ app.controller("AssignmentEditCtrl", ["$scope", "$window", "$state", "$statePara
...
+            $window.swal('Error', 'Queues could not be generated.', 'error');
+        });
+    }
+    
+    $scope.reloadAssignment();
+    $scope.reloadQueues();
+}])
+
+// obsolete for now, may need for more composition queue functionality
+app.controller("AssignmentQueueGenerateCtrl", ["$scope", "$window", "$state", "$stateParams", "Assignment", "Course", "Queues",
```
Comment (Sumukh) [inline server/static/js/admin/controllers.js]: Ah I see - this is for the future when we can assign to specific staff.

**Comment by Sumukh** [comment]
Made some comments - still have to test locally.

File: server/app/api.py (hunk 1 above)
Comment (alvinwan) [inline server/app/api.py]: Oh, this is for a theoretical "generate composition queues" based off of a set of user emails. I originally had users input a CSV, so I needed this check.

File: server/app/api.py (hunk 4 above)
Comment (alvinwan) [inline server/app/api.py]: Mhm. Should I keep the old CSV version down below? (queue.generate.html) Or should I go with a set of checkboxes?

**Comment by alvinwan** [comment]
@Sumukh Ready. Gif in initial post updated.

**Pull request merged**